contracts/HenkakuQuest.sol

```diff
@@ -2,25 +2,144 @@
+        require(
+            hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have an admin role to add Admin role"
+        );
+        _grantRole(ADMIN_ROLE, _to);
+    }
+
+    function addQuestCreationRole(address _to) public {
+        require(
+            hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have an admin role to add quest creation role"
+        );
+        _grantRole(QUEST_CREATION_ROLE, _to);
+    }
+
+    function addMemberRole(address _to) public {
+        require(
+            hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have an admin role to add quest creation role"
+        );
+        _grantRole(HENKAKU_MEMBER_ROLE, _to);
+    }
+
+    function isMember(address _address) internal view returns (bool) {
+        uint256 _balance = memberShipNFT.balanceOf(_address);
+        if (_balance <= 0) {
+            return false;
+        }
+
+        uint256 _tokenId = 0;
+        address _candidate;
+        bool _isMember = false;
+        while (!_isMember) {
+            _tokenId += 1;
+            _candidate = memberShipNFT.ownerOf(_tokenId);
+            _isMember = _candidate == _address;
+            console.log(_tokenId);
+            console.log(_candidate);
+        }
+        return memberShipNFT.isCommunityMember(_tokenId);
+    }
+
+    function closeQuest(uint256 _id) public {
+        require(
+            hasRole(QUEST_CREATION_ROLE, msg.sender) ||
+                hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have Quest creation role or admin role to edit a quest"
+        );
+        Quest memory _quest = quests[_id];
+        _quest.endedAt = block.timestamp;
+        _quest.endedAt = block.timestamp;
+        quests[_id] = _quest;
+    }
+
+    function update(uint256 _id, Quest memory _quest) public {
+        require(
+            hasRole(QUEST_CREATION_ROLE, msg.sender) ||
+                hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have Quest creation role or admin role to edit a quest"
+        );
+        quests[_id] = _quest;
+    }
 
     function save(Quest memory _quest) public {
-        quests.push(_quest);
+        require(
+            hasRole(QUEST_CREATION_ROLE, msg.sender) ||
+                hasRole(ADMIN_ROLE, msg.sender),
+            "You need to have Quest creation role or admin role to add a quest"
+        );
+        quests[id] = _quest;
+        id += 1;
     }
 
-    function getQuests() public view returns(Quest[] memory){
-        return quests;
+    function canReadQuest(address _address) public view returns (bool) {
+        return
+            hasRole(QUEST_CREATION_ROLE, _address) ||
+            hasRole(ADMIN_ROLE, _address) ||
+            hasRole(HENKAKU_MEMBER_ROLE, _address) ||
+            isMember(_address);
+    }
+
+    function getQuests() public view returns (Quest[] memory) {
+        require(canReadQuest(msg.sender), "You dont have permission to obtain");
```

Comment:
I usually don't put `msg.sender` in the view function so that the view function can be called without being connected to the wallet. Can you not just pass it via `getQuests(address)`? Having permission on the read function is a bit pointless as you are storing everything on blockchain anyways.

Reply:
FYI, token gated access is usually implemented off-chain using on-chain info, and I don't think you need to implement that on chain given it provides no protection.

Reply:
> the view function so that the view function can be called without being connected to the wallet.
I actually didn't know that we could call function without connecting the wallet.

>  I don't think you need to implement that on chain given it provides no protection.

Yes. but I wanted to make it difficult to access. that is the point.
ordinary people cannot reveal so that's actually enough.
since quest is not sensitive info so it will not be be problem even to be revealed.

but if some sort of encrypting methods exists, then it will be nice. 
do you know some methods to hold data on-chain and make it difficult to reveal ?

ideally, I also wanna hold sensitive info too if we can make secure.
I think that will be interesting research topic

Reply:
web4 lol
https://ethereum.stackexchange.com/a/13773

in current situation, workaround would be hashing data
https://yenthanh.medium.com/how-to-secure-sensitive-data-on-an-ethereum-smart-contract-c0823dde00c6